data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,104 @@
+import org.opensearch.dataprepper.peerforwarder.client.PeerForwarderClient;
+
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class PeerForwarder {
+    private final PeerForwarderClientFactory peerForwarderClientFactory;
+    private final PeerForwarderClient peerForwarderClient;
+    private HashRing hashRing;
+
+    public PeerForwarder(final PeerForwarderClientFactory peerForwarderClientFactory,
+                         final PeerForwarderClient peerForwarderClient) {
+        this.peerForwarderClientFactory = peerForwarderClientFactory;
+        this.peerForwarderClient = peerForwarderClient;
+    }
+
+    public List<Record<Event>> forwardRecords(final Collection<Record<Event>> records,
+                                              final Set<String> identificationKeys,
+                                              final String pluginId) {
+        // TODO: decide the default behaviour of Core Peer Forwarder and move the HashRing creation to constructor
+        hashRing = peerForwarderClientFactory.createHashRing();
+
+        final Map<String, List<Record<Event>>> groupedRecords = groupRecordsBasedOnIdentificationKeys(records, identificationKeys);
+
+        final List<Record<Event>> recordsToProcessLocally = new ArrayList<>();
+
+        for (final Map.Entry<String, List<Record<Event>>> entry: groupedRecords.entrySet()) {
+            final String destinationIp = entry.getKey();
+
+            if (isAddressDefinedLocally(destinationIp)) {
+                recordsToProcessLocally.addAll(entry.getValue());
+            }
+            else {
+                AggregatedHttpResponse httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
+                        destinationIp, pluginId);
+                if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
+                    recordsToProcessLocally.addAll(entry.getValue());
+                }
+            }
+        }
+        return recordsToProcessLocally;
+    }
+
+    private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(
+            final Collection<Record<Event>> records,
+            final Set<String> identificationKeys
+    ) {
+        final Map<String, List<Record<Event>>> groupedRecords = new HashMap<>();
+
+        // group records based on IP address calculated by HashRing
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            List<String> identificationKeyValues = new LinkedList<>();
+            for (final String identificationKey : identificationKeys) {
+                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
```

Comment:
We should be careful here. 

By default the toString() method will return a string the name of the class @ and a hexadecimal representation of the memory location. Are we sure the identifications keys will match across nodes with this behavior?

Reply:
Currently we only store String and primitive types, so the default implementation of toString for them should return string representation instead of hexadecimal representation.